### ISSUE TITLE
Compute V2: support for extended server information

### ISSUE THREAD

**Issue by ozerovandrei** [gophercloud/gophercloud#1160]
Currently OpenStack Compute service provides additional `OS-*` fields in the server struct that needs to be unmarshalled.
All of the extended fields are implemented in the separated extensions inside the Nova API. So it seems that it is better to implement them as separate packages in the Gophercloud.

For now Gophercloud implements only `extendedstatus` for the `OS-EXT-STS` fields (#659).
We need to also add support for the following info:

- [x] `OS-EXT-STS` Extended Status Extension.
- [x] `OS-DCF` Disk Management Extension.
- [x] `OS-EXT-AZ` Extended Availability Zone support. 
- [x] `OS-EXT-SRV-ATTR` partial Extended Server Attributes support. (#1163)
- [x] `OS-EXT-SRV-ATTR` full (depends on #1115 microversioning) Extended Server Attributes support (#1501)
- [x] `OS-SRV-USG` Usage (launched_at and terminated_at) support. (#1161)

**Comment by ozerovandrei** [comment]
It seems that `OS-DCF` extension was merged into main server struct:
https://github.com/openstack/nova/blob/stable/queens/nova/api/openstack/compute/views/servers.py#L150

So what's better way to implement that field?
Do I need to separate it in the different extension package or it is completely fine to add it into the `Server` struct?

**Comment by jtopjian** [comment]
@ozerovandrei Thank you for spotting the missing features :)

The disk config extension is already defined here:

https://github.com/gophercloud/gophercloud/tree/master/openstack/compute/v2/extensions/diskconfig

If both the request and response are part of the base API, we could add it to the base server `CreateOpts` and `Server` struct.

Availability Zones are already defined here: https://github.com/gophercloud/gophercloud/blob/master/openstack/compute/v2/extensions/availabilityzones/results.go

It looks like `OS-EXT-AZ` is indeed missing from Gophercloud. I can review this shortly.

For `OS-EXT-SRV-ATTR`: since it requires a microversion, I'm hesitant to merge it right now. There should be a solution once #1115 is merged.

**Comment by ozerovandrei** [comment]
@jtopjian oh, I missed those two packages for some reason.

Well, I personally wanted to propose a separate package for the `OS-DCF` so the main `compute/v2/servers` package won't become a huge mess (at least for now).
But it seems that Nova extensions are in slow progress of being merged into single module inside the Nova API server. So I guess i some future we will actually need to merge all of those stuff into one package.
I think that it's better not to move implemented methods from one place to other inside the library, at least for now. So let `compute/v2/extensions/diskconfig` stay where it is now.
 
Some note about microversions and `OS-EXT-SRV-ATTR`: users need to pass microversion only to get all of the fields. They still can get `hypervisor_hostname` and `instance_name` without specifying it.

**Comment by jtopjian** [comment]
> Well, I personally wanted to propose a separate package for the OS-DCF so the main compute/v2/servers package won't become a huge mess (at least for now).

I agree with this. I raised this topic last year here: https://github.com/gophercloud/gophercloud/issues/425

A year later, I think the current directory structure is still valid and doesn't need to be changed at all. When possible, we can continue having separate packages for different Nova/Compute API actions. When it makes sense to amend the Server directly, we'll do that.

> I think that it's better not to move implemented methods from one place to other inside the library, at least for now. So let compute/v2/extensions/diskconfig stay where it is now.

Agreed. Makes sense to me.

> Some note about microversions and OS-EXT-SRV-ATTR: users need to pass microversion only to get all of the fields. They still can get hypervisor_hostname and instance_name without specifying it.

From reading the Nova code (I didn't test this), it looked like that from microversion 2.4+ the fields are all available with the `OS-EXT-SRV-ATTR:` prefix. I might have misinterpreted this.

So it looks like there are 3 different behaviors depending on if no microversion was specified, a 2.3 microversion was specified, and a 2.4+ microversion was specified.

If you wanted to work on the non-microversion behavior, go for it. We can look at the rest later if that makes sense.

**Comment by ozerovandrei** [comment]
> If you wanted to work on the non-microversion behavior, go for it. We can look at the rest later if that makes sense.

This actually makes sense.
So can I leave only `instance_name`, `host` and `hypervisor_hostname` for now (I mean, in the current PR #1163 )? I can also add some inline comments about future fields and microversion support.

In that case we can save this issue for later and add all other `OS-EXT-SRV-ATTR` fields when there will be a clear way of how to deal with fields that are dependent on the microversions.

**Comment by jtopjian** [comment]
That all sounds good to me :)

**Comment by jtopjian** [comment]
:tada: 

Thanks, @ozerovandrei!
